### PR TITLE
Add dynamic template support to Twilio Sendgrid binding

### PR DESCRIPTION
**Pull request by yannart**
# Description

Adding support for dynamic templates. This includes allowing to provide a template id and template JSON data.
Supports these additional optional metadata (in configuration or request):
* dynamicTemplateId: the template id. Eg. "d-123456"
* dynamicTemplateData: the JSON passed to Sendgrid to populate the template. Eg. '{"customerName":"John Smith"}'

## Issue reference

Please reference the issue this PR will close: #2412

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo. Issue: dapr/docs#3421 PR: https://github.com/dapr/docs/pull/3422

### CODE REVIEW

File: bindings/twilio/sendgrid/sendgrid.go
Comment (ItalyPaleAle) [inline bindings/twilio/sendgrid/sendgrid.go:273]:
```suggestion
		return fmt.Errorf("error from SendGrid binding, dynamic template data is not valid JSON: %w", err)
```

File: bindings/twilio/sendgrid/sendgrid_test.go
Comment (ItalyPaleAle) [inline bindings/twilio/sendgrid/sendgrid_test.go:76]:
```suggestion
		assert.NoError(t, err)
```

File: bindings/twilio/sendgrid/sendgrid_test.go
Comment (ItalyPaleAle) [inline bindings/twilio/sendgrid/sendgrid_test.go:81]:
```suggestion
		assert.Error(t, err)
```

File: bindings/twilio/sendgrid/sendgrid.go
Comment (ItalyPaleAle) [inline bindings/twilio/sendgrid/sendgrid.go:231]: Are templateID and templateData mutually exclusive?

File: bindings/twilio/sendgrid/sendgrid.go
Comment (ItalyPaleAle) [inline bindings/twilio/sendgrid/sendgrid.go:197]: Because `sg.metadata` is parsed in the Init method, you should be able to do this un-marshaling only once and cache the result. This should improve perf quite a bit

**Review by ItalyPaleAle — COMMENTED** [review]
Thanks for this PR! Left a few comments and suggestions

**Comment by yannart** [comment]
@ItalyPaleAle Thanks for your comments and suggestions. I have attended all of them and did a local run to verify and test using my Sendgrid account.

**Review by ItalyPaleAle — DISMISSED** [review]

**Comment by ItalyPaleAle** [comment]
Thank you, LGTM!

Sadly, because we are into the endgame for Dapr 1.11 already, this will need to land in Dapr 1.12.

PS: Would you be interested in contributing certification tests for SendGrid? This way we can also mark the component as stable

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by yann-nicolas** [comment]
> Thank you, LGTM!
> 
> Sadly, because we are into the endgame for Dapr 1.11 already, this will need to land in Dapr 1.12.
> 
> PS: Would you be interested in contributing certification tests for SendGrid? This way we can also mark the component as stable

Thanks for you PR approval. Question on merge process, need to wait for approval from "approvers-components-contrib" and then I merge?

On the PS: Yes, I was actually thinking about contributing to the Conformance tests so I am interested and should be able to work on it next weeks. However will need guidance as not sure how you test the bindings with 3rd parties (eg. do you need to have your own Sendgrid account or do we mock the API?).

**Comment by ItalyPaleAle** [comment]
I can merge this, was just waiting fr tests to pass.

> On the PS: Yes, I was actually thinking about contributing to the Conformance tests so I am interested and should be able to work on it next weeks. However will need guidance as not sure how you test the bindings with 3rd parties (eg. do you need to have your own Sendgrid account or do we mock the API?).

For a binding like this, I'd recommend a Certification test, which is specific to the binding. Conformance tests are not particularly useful here, as this binding is fairly unique. You can see an example of a certification test for a binding here: https://github.com/dapr/components-contrib/tree/master/tests/certification/bindings/cron

We generally prefer tests to run against a real service. If that's not possible, and in this case I can understand why that would be a problem, mocking may be an option too.

**Comment by yannart** [comment]
> I can merge this, was just waiting fr tests to pass.
Perfect.

> For a binding like this, I'd recommend a Certification test, which is specific to the binding. Conformance tests are not particularly useful here, as this binding is fairly unique. You can see an example of a certification test for a binding here: https://github.com/dapr/components-contrib/tree/master/tests/certification/bindings/cron
> We generally prefer tests to run against a real service. If that's not possible, and in this case I can understand why that would be a problem, mocking may be an option too.
I'll look at it and create issue.

**Pull request merged**
